src/storage/storage.cpp

```diff
@@ -184,643 +182,670 @@ Storage::ReturnCode Storage::Run(int max_wait)
+            memory_ptr, DataLayout::GEOMETRIES_REV_WEIGHT_LIST);
+        BOOST_ASSERT(geometry_node_lists_count ==
+                     layout_ptr->num_entries[DataLayout::GEOMETRIES_REV_WEIGHT_LIST]);
+        geometry_input_file.readInto(geometries_rev_weight_list_ptr, geometry_node_lists_count);
+    }
 
-    for (auto i = 0u; i < number_of_core_markers; ++i)
     {
-        BOOST_ASSERT(unpacked_core_markers[i] == 0 || unpacked_core_markers[i] == 1);
+        io::File geometry_datasource_file(config.datasource_indexes_path);
+        const auto number_of_compressed_datasources = geometry_datasource_file.readElementCount64();
 
-        if (unpacked_core_markers[i] == 1)
+        // load datasource information (if it exists)
+        const auto datasources_list_ptr =
+            layout_ptr->GetBlockPtr<uint8_t, true>(memory_ptr, DataLayout::DATASOURCES_LIST);
+        if (number_of_compressed_datasources > 0)
         {
-            const unsigned bucket = i / 32;
-            const unsigned offset = i % 32;
-            const unsigned value = [&] {
-                unsigned return_value = 0;
-                if (0 != offset)
-                {
-                    return_value = core_marker_ptr[bucket];
-                }
-                return return_value;
-            }();
-
-            core_marker_ptr[bucket] = (value | (1u << offset));
+            io::readDatasourceIndexes(
+                geometry_datasource_file, datasources_list_ptr, number_of_compressed_datasources);
         }
     }
 
-    // load the nodes of the search graph
-    QueryGraph::NodeArrayEntry *graph_node_list_ptr =
-        shared_layout_ptr->GetBlockPtr<QueryGraph::NodeArrayEntry, true>(
-            shared_memory_ptr, SharedDataLayout::GRAPH_NODE_LIST);
+    {
+        /* Load names */
+        io::File datasource_names_file(config.datasource_names_path);
 
-    // load the edges of the search graph
-    QueryGraph::EdgeArrayEntry *graph_edge_list_ptr =
-        shared_layout_ptr->GetBlockPtr<QueryGraph::EdgeArrayEntry, true>(
-            shared_memory_ptr, SharedDataLayout::GRAPH_EDGE_LIST);
+        const auto datasource_names_data = io::readDatasourceNames(datasource_names_file);
 
-    io::readHSGR(hsgr_input_stream,
-                 graph_node_list_ptr,
-                 hsgr_header.number_of_nodes,
-                 graph_edge_list_ptr,
-                 hsgr_header.number_of_edges);
-    hsgr_input_stream.close();
+        // load datasource name information (if it exists)
+        const auto datasource_name_data_ptr =
+            layout_ptr->GetBlockPtr<char, true>(memory_ptr, DataLayout::DATASOURCE_NAME_DATA);
+        if (layout_ptr->GetBlockSize(DataLayout::DATASOURCE_NAME_DATA) > 0)
+        {
+            std::copy(datasource_names_data.names.begin(),
+                      datasource_names_data.names.end(),
+                      datasource_name_data_ptr);
```

Comment:
Please check all copy's to make sure you assert for end - begin == size of the rout range.

Maybe even copy_n.